src/sage/algebras/quatalg/quaternion_algebra.py

```diff
@@ -2477,6 +2500,43 @@ def basis_matrix(self):
         C, d = B._clear_denom()
         return C.hermite_form() / d
 
+    def reduced_basis(self):
+        r"""
+        Let `I` = ``self`` be a rank 4 quaternion ideal in a definite quaternion algebra.
+        This function returns an LLL reduced basis of I.
+
+        OUTPUT:
+
+        - A tuple of four elements in I forming an LLL reduced basis of I as a lattice
+
+        EXAMPLES::
+
+            sage: B = BrandtModule(2,37); I = B.right_ideals()[0]
+            sage: I
+            Fractional ideal (2 + 2*i + 2*j + 2*k, 4*i + 108*k, 4*j + 44*k, 148*k)
+            sage: I.reduced_basis()
+            (2 + 2*i + 2*j + 2*k, 4, -2 - 2*i - 14*j + 14*k, -16*i + 12*k)
+            sage: l = I.reduced_basis()
+            sage: assert all(l[i].reduced_norm() <= l[i+1].reduced_norm() for i in range(len(l) - 1))
+
+            sage: B = QuaternionAlgebra(next_prime(2**50))
+            sage: O = B.maximal_order()
+            sage: i,j,k = B.gens()
+            sage: alpha = 1/2 - 1/2*i + 3/2*j - 7/2*k
+            sage: I = O*alpha + O*3089622859
+            sage: I.reduced_basis()[0]
+            1/2*i + j + 5/2*k
+        """
+        if len(self.basis()) < 4:
+            raise ValueError("basis must have rank 4")
+
+        if not self.quaternion_algebra().is_definite():
+            if not self.quadratic_form().is_positive_definite():
+                raise TypeError("The quaternion algebra must be definite")
```

Comment:
Aren't these two tests equivalent? Were they duplicated by accident?

Reply:
Yes, these two should be equivalent - I believe the latter test was originally put in place for the method to also work for "lattices" of non-maximal rank, until it came up that `.gram_matrix()` needs a full rank lattice anyway. I'll remove the second test.